[SYCL][ESIMD] Add mutator for test input data

vasilytric opened the pull request:
Provides the following things:
- base options to mutate value.
- function that applies provided mutator to each value for provided container

vasilytric committed:
[SYCL][ESIMD] Add mutator for test input data

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,87 @@
+
+namespace esimd_test::api::functional {
+
+// Provides base functions to mutate value.
```
```suggestion
// In some test cases it’s possible to pass a reference data as the input data
// directly without any modification. For example, if we check some copy
// constructor or memory move operation it’s OK to have any data values as the
// input ones. But in most cases we need to consider the possibility of UB for
// C++ operations due to modification of input values.
//
// Mutation mechanism covers such requirement.
// The mutator namespace is intended to store such mutators alongside with
// the generic ones provided below.
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/789/commits/7cc2f674c94df0416edeeb133af0de198829cfb6#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R25-R33)

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,87 @@
+// Provides base functions to mutate value.
+namespace mutator {
+
+template <typename T> class Add {
```
```suggestion
// Replace specific reference values to the bigger ones, so we can safely
// substract `value_for_mutation` later
template <typename T> class Add {
```
That's what we discussed earlier, but the logic below doesn't fit such description, because:

- instead of disabling UB for "substraction" operation by changing the set of input values from `[min..max]` to `[min+val..max]`, by only handling the `min` border
- we are just adding `val` for every reference value, effectively disabling some corner case checks (`denorm_min` for example)

Please verify the logic for mutators. Feel free to rename them if the naming suggested feels confusing - for example `mutator::for_substraction` instead of `mutator::add`

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/789/commits/7cc2f674c94df0416edeeb133af0de198829cfb6#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74L34-L40)

Also mutators has been renamed

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,87 @@
+public:
+  Add(T value_for_mutation) : m_value(value_for_mutation) {}
+
+  void operator()(T &value_for_mutation) {
```
Please consider rename one of these `value_for_mutation` parameters.
It would be fine to have just `value` in both cases - because for each of these cases it would have the same meaning "input value for this function"
But given the `_for_mutation` suffix we expect it to have some additional meaning. Either value to use as a base for mutation. Or a value to check using such base.

vasilytric replied:
`value_for_mutation` has been renamed into `val` [here](https://github.com/intel/llvm-test-suite/pull/789/commits/7cc2f674c94df0416edeeb133af0de198829cfb6#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R60) and in all cases
I can't use `value` name because we have another one `value` that uses for obtain lowest, max, etc. values

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,87 @@
+
+  void operator()(T &value_for_mutation) {
+    if constexpr (std::is_signed_v<T>) {
+      if (value<T>::max() - m_value <= value_for_mutation) {
```
Please note that you suggest `m_value` is positive here, so we would need assertion in constructor to ensure such requirement.

vasilytric replied:
`assert` has been added [here](https://github.com/intel/llvm-test-suite/pull/789/commits/7cc2f674c94df0416edeeb133af0de198829cfb6#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R60) and for all mutators

vasilytric committed:
[SYCL][ESIMD] Update logic for mutator functors and update comments

vasilytric committed:
[SYCL][ESIMD] Add necessary includes

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,94 @@
+
+// Replace specific reference values to the divided ones, so we can safely
+// multiply to `val` later.
+template <typename T> class For_division {
```
```suggestion
// Replace specific reference values to the divided ones, so we can safely
// multiply to `val` later.
template <typename T> class For_multiplication {
```
given the description above. Looks like the logic should be fixed therefore

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R93)

yuriykoch replied:
Please, update the logic. Let's for example have `using T = int`, `m_value == 2` and `val == value<int>::max()` in `operator()`.  
Looks like we don't need to touch `denorm_min` at all.

vasilytric replied:
The logic has been updated [here](https://github.com/intel/llvm-test-suite/pull/789/commits/41df5e531f17caed6253334af281f97f021058d3#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74L101-R106)

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,94 @@
+// Applies provided mutator to each value for provided container.
+template <typename T, typename MutatorT>
+void mutate(std::vector<T> &input_vector, MutatorT &&mutator) {
+  std::for_each(input_vector.begin(), input_vector.end(), mutator);
```
```suggestion
inline void mutate(std::vector<T> &input_vector, MutatorT &&mutator) {
  std::for_each(input_vector.begin(), input_vector.end(), mutator);
```
Nit

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R111)

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,94 @@
+  For_addition(T val) : m_value((assert(val > 0 && "Invalid value."), val)) {}
+
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T> || std::is_same_v<T, sycl::half>) {
```
Good catch. Let's have a type_trait for it maybe?

vasilytric replied:
How we can name it? `is_sycl_signed`?

yuriykoch replied:
Looks good

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R49) and the aliases has defined [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-7b267ff200ef7dbfe5e2b1fe912dcbc7a8622f4d14a995f2f683282cff5e1b49R32-R38)

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,94 @@
+    if constexpr (std::is_signed_v<T>) {
+      if (val < value<T>::lowest() + m_value) {
+        val = value<T>::lowest() + m_value;
+      }
```
```suggestion
      const T lower_border = value<T>::lowest() + m_value;
      if (val < lower_border) {
        val = lower_border;
      }
```

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R57-R59)

yuriykoch commented on SYCL/ESIMD/esimd_test_utils.hpp:
```diff
@@ -6,6 +6,8 @@
 
+#pragma once
+
 #include <CL/sycl.hpp>
```
```suggestion
#include <sycl/sycl.hpp>
```
While you a here :) Great catch with pragma BTW

vasilytric replied:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-c73d28d272bf069fdc585bd82fc0fcbcba32f9b98a256e21781cfedc5080d0e5R11)

vasilytric committed:
[SYCL][ESIMD] Small updates

- Add handlers for infinity values cases
- Rename mutator
- Replace "#include <CL/sycl.hpp>" with "#include <sycl/sycl.hpp>"

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,115 @@
+  void operator()(T &val) {
+    // Floating point variable with infinity value shouldn't be changed.
+    if constexpr (type_traits::is_sycl_signed_v<T>) {
+      if constexpr (type_traits::is_sycl_floating_point_v<T>) {
+        // sycl::half will be converted to float
+        if (std::isinf(val)) {
+          return;
+        }
+      }
+
+      const T lower_border = value<T>::lowest() + m_value;
+      if (val < lower_border) {
+        val = lower_border;
+      }
```
```suggestion
      const T lower_border = value<T>::lowest() + m_value;
      if (val < lower_border) {
        // Validate only the negative infinity case, as the NaN will not enter this branch
        if constexpr (type_traits::is_sycl_floating_point_v<T>) {
          // sycl::half will be converted to float
          if (std::isinf(val)) {
            return;
          }
        }
	// we need to update value to avoid UB during addition
        val = lower_border;
      }
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/789/commits/41df5e531f17caed6253334af281f97f021058d3#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R47-R64) and in other mutators

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,115 @@
+namespace mutator {
+
+// Replace specific reference values to the bigger ones, so we can safely
+// substract `val` later.
```
```suggestion
// substract `m_value` later.
```
We have multiple `val` here, but single `m_value`

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/789/commits/41df5e531f17caed6253334af281f97f021058d3#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R39) and in other cases

yuriykoch commented on SYCL/ESIMD/api/functional/type_traits.hpp:
```diff
@@ -27,6 +29,13 @@ template <typename T>
 
+template <typename T>
+using is_sycl_signed =
+    std::bool_constant<std::is_signed_v<T> || std::is_same_v<T, sycl::half>>;
```
```suggestion
    std::bool_constant<std::is_signed_v<T> || is_sycl_floating_point_v<T>>;
```
Nit.
Just as a suggestion for DRY

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/789/commits/41df5e531f17caed6253334af281f97f021058d3#diff-7b267ff200ef7dbfe5e2b1fe912dcbc7a8622f4d14a995f2f683282cff5e1b49R34)

vasilytric committed:
[SYCL][ESIMD] Update functions logic and comments

yuriykoch commented on SYCL/ESIMD/api/functional/mutator.hpp:
```diff
@@ -0,0 +1,119 @@
+    if (val > upper_border) {
+      // we need to update value to avoid UB during multiplication.
+      val = upper_border;
+    }
```
Please don't forget about `min`. We might multiply negative values also.
Also we don't need to change the `inf` values, because `inf * value` gives `inf`

vasilytric replied:
Good idea! :)
I've done it in this commit: https://github.com/intel/llvm-test-suite/pull/789/commits/c9ec74dadeec641b3eddde8795b4829fe739451b
The deletion will be skipped in case `val == inf` or `val == -inf`
I've done it in this commit: 33412947b2c925289c56bce4e72917d336ab2f3d

vasilytric committed:
[SYCL][ESIMD] Update case with negative numbers for "For_multiplication"

vasilytric committed:
[SYCL][ESIMD] Skip deletion in case that current element is infinity

vasilytric commented:
@sndmitriev Would you mind to take a look?

vasilytric commented:
@kbobrovs Sergey have approved this PR would you mind to merge this PR?
